File: spring-boot-oauth2-auth-server/src/main/java/com/bittokazi/oauth2/auth/server/app/controllers/RoleController.java
package com.bittokazi.oauth2.auth.server.app.controllers;

import com.bittokazi.oauth2.auth.server.app.models.tenant.Role;
import com.bittokazi.oauth2.auth.server.app.services.role.RoleService;
import com.bittokazi.oauth2.auth.server.app.services.role.RoleServiceImpl;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;


/**
 * @author Bitto Kazi
 */

@RestController
@RequestMapping("/api")
public class RoleController {

	private RoleService roleService;

	public RoleController(RoleServiceImpl roleServiceImpl) {
		this.roleService = roleServiceImpl;
	}

	@PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
	@GetMapping("/roles")
	public Object getRoles(@RequestParam(value = "page", required = false, defaultValue = "0") Integer page,
						   @RequestParam(value = "count", required = false, defaultValue = "10") Integer count) {
		return roleService.getRoles(page, count);
	}

	@PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
	@GetMapping("/roles/all")
	public Object getAllRoles() {
		return roleService.getAllRoles();
	}

	@PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
	@PostMapping("/roles")
	public Object addRole(@RequestBody Role role, HttpServletRequest httpServletRequest,
						  HttpServletResponse httpServletResponse) {

		return roleService.addRole(role, httpServletRequest, httpServletResponse);
	}

	@PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
	@GetMapping("/roles/{id}")
	public Object getRole(@PathVariable String id) {
		return roleService.getRole(id);
	}

	@PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
	@PutMapping("/roles/{id}")
	public Object updateRole(@PathVariable String id, @RequestBody Role role,
									HttpServletResponse response) {
		return roleService.updateRole(role, id, response);
	}

	@PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
	@PostMapping("/roles/search/name")
	public Object getRoleByName(@RequestBody Role role, HttpServletRequest httpServletRequest,
						  HttpServletResponse httpServletResponse) {
		return roleService.getRoleByName(role);
	}
}

File: spring-boot-oauth2-auth-server/src/main/kotlin/com/bittokazi/oauth2/auth/server/app/controllers/RoleController.kt
package com.bittokazi.oauth2.auth.server.app.controllers

import com.bittokazi.oauth2.auth.server.app.models.tenant.Role
import com.bittokazi.oauth2.auth.server.app.services.role.RoleService
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

/**
 * @author Bitto Kazi
 */
@RestController
@RequestMapping("/api")
class RoleController(
    private val roleService: RoleService
) {

    @PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
    @GetMapping("/roles")
    fun getRoles(
        @RequestParam(value = "page", required = false, defaultValue = "0") page: Int,
        @RequestParam(value = "count", required = false, defaultValue = "10") count: Int
    ): Any = roleService.getRoles(page, count)

    @GetMapping("/roles/all")
    @PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
    fun allRoles(): Any = roleService.allRoles()

    @PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
    @PostMapping("/roles")
    fun addRole(
        @RequestBody role: Role,
        httpServletRequest: HttpServletRequest,
        httpServletResponse: HttpServletResponse
    ): Any = roleService.addRole(role, httpServletRequest, httpServletResponse)

    @PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
    @GetMapping("/roles/{id}")
    fun getRole(
        @PathVariable id: String
    ): Any = roleService.getRole(id)

    @PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
    @PutMapping("/roles/{id}")
    fun updateRole(
        @PathVariable id: String,
        @RequestBody role: Role,
        response: HttpServletResponse
    ): Any = roleService.updateRole(role, id, response)

    @PreAuthorize("hasAuthority('SCOPE_role:all') or hasAuthority('SCOPE_SUPER_ADMIN')")
    @PostMapping("/roles/search/name")
    fun getRoleByName(
        @RequestBody role: Role,
        httpServletRequest: HttpServletRequest,
        httpServletResponse: HttpServletResponse
    ): Any = roleService.getRoleByName(role)
}
